### PR TITLE
feat(smithy-client): add handler cache

### PR DESCRIPTION
**Pull request by kuhe**
https://github.com/aws/aws-sdk-js-v3/issues/6423

This adds a cache for the handler functions resolved from calls to 

```ts
command.resolveMiddleware(client.middlewareStack, client.config, (request) options)
```

when options is undefined. This occurs in the request path and can take 1-2ms, but is redundant for any client/command pair in most cases. 

Despite being cached, the handler functions has a reference to the client config, so any changes there will still be visible to the middleware within the resolved handler. ~Modifying a client's middlewareStack invalidates the cache.~

The cache is opt-in and also unique per client, so users retain control that way.

### CODE REVIEW

File: packages/smithy-client/src/client.ts [hunk 1]
```diff
@@ -87,6 +113,7 @@ export class Client<
   }
 
   destroy() {
-    if (this.config.requestHandler.destroy) this.config.requestHandler.destroy();
+    this.config.requestHandler.destroy?.();
+    clientHandlers.delete(this);
```
Comment (kuhe) [inline packages/smithy-client/src/client.ts:117]: if the concern is affecting the shape of the class, we can probably use `private handler?` i.e. optionality in the property, for backwards compatibility.

File: packages/smithy-client/src/client.ts [hunk 2]
```diff
@@ -44,11 +45,21 @@ export class Client<
...
+      delete this.handlers;
+    },
+  });
+  /**
+   * May be used to cache the resolved handler function for a Command class.
+   */
+  private handlers?: WeakMap<Function, Handler<any, any>> | undefined;
+  private configRef?: ResolvedClientConfiguration | undefined;
+
+  constructor(public readonly config: ResolvedClientConfiguration) {
```
Comment (milesziemer) [inline packages/smithy-client/src/client.ts:59]: Maybe just me not being familiar with this syntax, but it looks like `this.configRef` is what's assigned in the ctor, not `this.config`. Is `this.config` still accessible even though it isn't declared or assigned to? I.e. in the `destroy` method.

Comment (kuhe) [reply packages/smithy-client/src/client.ts:59]: this is shorthand, these are equivalent

```ts
public constructor(public prop: Type) {}

public constructor(prop: Type) {
  this.prop = prop;
}
```

configRef is for detecting reference changes to the client's config property, but this whole PR is kind of iffy. Probably not worth putting change detectors on so much stuff just to save 1ms of handler resolution time, and it's leaky.

**Comment by kuhe** [comment]
putting this into draft while I reevaluate whether this change is worth it or not

File: packages/smithy-client/src/client.ts (hunk 2 above)
Comment (kuhe) [reply packages/smithy-client/src/client.ts]: no longer applicable in latest revision

**Review by milesziemer — APPROVED** [review]

**Review by siddsriv — APPROVED** [review]

**Pull request merged**